File: packages/go-agentenkit/compaction_test.go
package agentenkit_test

import (
	"context"
	"encoding/json"
	"fmt"
	"strings"
	"sync"
	"testing"

	"github.com/zendev-sh/goai/provider"

	agentenkit "github.com/eadwinCode/agentic-kit/packages/go-agentenkit"
	"github.com/eadwinCode/agentic-kit/packages/go-agentenkit/core"
	"github.com/eadwinCode/agentic-kit/packages/go-agentenkit/ports"
)

// Workstream E: a summary covers what it summarized, so the prompt stops
// carrying it, a thread compacts once per time it outgrows the trigger, and
// the kept tail starts at a user turn. The same cases run in the TS package
// (test/compaction.test.ts).

// summarizingModel answers every streamed call "ok" and every summary
// request "summary", and keeps the prompts it streamed.
type summarizingModel struct {
	mu      sync.Mutex
	prompts [][]provider.Message
}

func (m *summarizingModel) ModelID() string { return "mock-summarizing" }

func (m *summarizingModel) DoGenerate(context.Context, provider.GenerateParams) (*provider.GenerateResult, error) {
	return &provider.GenerateResult{Text: "summary", FinishReason: provider.FinishStop,
		Usage: provider.Usage{InputTokens: 100, OutputTokens: 10}}, nil
}

func (m *summarizingModel) DoStream(_ context.Context, p provider.GenerateParams) (*provider.StreamResult, error) {
	m.mu.Lock()
	m.prompts = append(m.prompts, p.Messages)
	m.mu.Unlock()
	ch := make(chan provider.StreamChunk, 2)
	ch <- provider.StreamChunk{Type: provider.ChunkText, Text: "ok"}
	ch <- provider.StreamChunk{Type: provider.ChunkFinish, FinishReason: provider.FinishStop, Usage: provider.Usage{InputTokens: 10, OutputTokens: 5}}
	close(ch)
	return &provider.StreamResult{Stream: ch}, nil
}

func (m *summarizingModel) lastPrompt() string {
	m.mu.Lock()
	defer m.mu.Unlock()
	b, _ := json.Marshal(m.prompts[len(m.prompts)-1])
	return string(b)
}

// compactingRuntime: a 1,800-token budget, compacting past 1,440, keeping a
// 450-token tail. Each prompt below is about 300 tokens.
func compactingRuntime(t *testing.T) (*harness, *agentenkit.AgentHandle, *summarizingModel) {
	model := &summarizingModel{}
	h := makeRuntimeOpts(t, scripted(step{text: "unused"}), func(o *agentenkit.RuntimeOptions) {
		o.ResolveModel = func(string) (agentenkit.ResolvedModel, error) {
			return agentenkit.ResolvedModel{Instance: func() provider.LanguageModel { return model }, ContextWindow: 128_000}, nil
		}
	}, func(c *agentenkit.AgentConfig) {
		c.ContextCeilingTokens = 2_000
		c.ContextOutputReserveTokens = 200
		c.CompactionModel = "gpt-4o"
		c.PromptCaching = false
	})
	chat := h.rt.CreateStreamTextAgent(agentenkit.StreamTextAgentSpec{Name: "chat", Model: "gpt-4o"})
	return h, chat, model
}

func runPrompt(t *testing.T, h *harness, chat *agentenkit.AgentHandle, threadID string, i int) agentenkit.RunResult {
	t.Helper()
	ran := h.run(t, chat, agentenkit.RunInput{ThreadID: threadID, Prompt: fmt.Sprintf("run-%d %s", i, strings.Repeat("x", 1_200))})
	h.handleNext(t)
	return ran
}

func summaries(t *testing.T, h *harness, threadID string) int {
	t.Helper()
	msgs, _ := h.storage.Messages().List(h.ctx, threadID, agentenkit.MainAgent, agentenkit.StorageContext{})
	n := 0
	for _, m := range msgs {
		if strings.Contains(string(m.Content), core.ContextSummaryType) {
			n++
		}
	}
	return n
}

func TestCompaction_AThreadCompactsOnceEachTimeItGrowsPastTheTrigger(t *testing.T) {
	h, chat, model := compactingRuntime(t)
	first := runPrompt(t, h, chat, "", 1)
	for i := 2; i <= 6; i++ {
		runPrompt(t, h, chat, first.ThreadID, i)
	}
	mustEqual(t, summaries(t, h, first.ThreadID), 1, "compacted once, not on every run after")
	if strings.Contains(model.lastPrompt(), "run-1 ") {
		t.Fatal("a summarized turn is not sent again")
	}
	if !strings.Contains(model.lastPrompt(), "run-6 ") {
		t.Fatal("the new turn is sent")
	}
}

func TestCompaction_TheSummaryIsBilledToTheRunItServed(t *testing.T) {
	h, chat, _ := compactingRuntime(t)
	first := runPrompt(t, h, chat, "", 1)
	var compacting agentenkit.RunResult
	for i := 2; i <= 5; i++ {
		compacting = runPrompt(t, h, chat, first.ThreadID, i)
	}
	for _, r := range h.storage.UsageRows() {
		if r.Kind == agentenkit.KindCompaction {
			mustEqual(t, r.RunID, compacting.RunID, "billed to the run it served")
			return
		}
	}
	t.Fatal("no compaction call recorded")
}

func TestCompaction_TheKeptTailStartsAtAUserTurn(t *testing.T) {
	h, _, _ := compactingRuntime(t)
	deps := h.rt.Ports(nil)
	th, _ := h.storage.Threads().Create(h.ctx, ports.ThreadInit{}, agentenkit.StorageContext{})
	add := func(role ports.MessageRole, content json.RawMessage) {
		if _, err := h.storage.Messages().Append(h.ctx, th.ID, ports.NewMessage{Role: role, Content: content}, agentenkit.StorageContext{}); err != nil {
			t.Fatal(err)
		}
	}
	big := strings.Repeat("y", 2_400)
	add(ports.RoleUser, core.TextContent("first "+big))
	add(ports.RoleUser, core.TextContent("second "+big))
	add(ports.RoleAssistant, core.PartsContent([]core.ContentPart{{Type: "tool-call", ToolCallID: "c1", ToolName: "lookup", Args: json.RawMessage(`{"q":"` + strings.Repeat("q", 300) + `"}`)}}))
	// A result that fits the tail on its own, with its call just outside it.
	add(ports.RoleTool, core.ToolResultContent("c1", "lookup", strings.Repeat("z", 1_300)))
	add(ports.RoleAssistant, core.TextContent("looked it up"))
	add(ports.RoleUser, core.TextContent("third"))
	out, err := agentenkit.CompactContext(h.ctx, deps, th.ID, "gpt-4o", core.CompactOptions{})
	if err != nil {
		t.Fatal(err)
	}
	mustEqual(t, out[0].Role, ports.RoleSystem, "the summary leads")
	mustEqual(t, out[1].Role, ports.RoleUser, "the tail starts at a user turn")
	mustEqual(t, len(out), 2, "summary + the last user turn")
}

func TestCompaction_ASummaryFromBeforeTheCoverMarkIsLeftOut(t *testing.T) {
	msg := func(id string, role ports.MessageRole, content json.RawMessage) ports.MessageDTO {
		return ports.MessageDTO{ID: id, Role: role, Content: content}
	}
	u1, a1, u2 := msg("u1", ports.RoleUser, core.TextContent("one")), msg("a1", ports.RoleAssistant, core.TextContent("two")), msg("u2", ports.RoleUser, core.TextContent("three"))
	legacy := msg("s0", ports.RoleSystem, json.RawMessage(`{"type":"CONTEXT_SUMMARY","text":"old"}`))
	marked := msg("s1", ports.RoleSystem, core.ContextSummaryContent("new", "a1"))

	ids := func(ms []ports.MessageDTO) []string {
		var out []string
		for _, m := range ms {
			out = append(out, m.ID)
		}
		return out
	}
	mustStrings(t, ids(core.PromptHistory([]ports.MessageDTO{u1, a1, legacy, u2})), []string{"u1", "a1", "u2"}, "a legacy summary covers nothing known")
	mustStrings(t, ids(core.PromptHistory([]ports.MessageDTO{u1, a1, u2, marked})), []string{"s1", "u2"}, "the summary, then what came after what it covers")
}

File: packages/agentenkit/test/compaction.test.ts
import { describe, expect, it } from 'bun:test';
import { simulateReadableStream } from 'ai';
import type { LanguageModelV1StreamPart } from '@ai-sdk/provider';
import { MockLanguageModelV1 } from 'ai/test';
import { setupAgentCore } from '../src/runtime.js';
import { MemoryAdminStore } from '../src/admin/memory.js';
import { MemoryBus, MemoryKv, MemoryQueue, MemoryStorage } from '../src/adapters/memory.js';
import { bindStorage } from '../src/core/state.js';
import { compactContext } from '../src/core/context.js';
import { promptHistory } from '../src/core/messages.js';
import { resolveConfig } from '../src/core/types.js';
import type { RuntimePorts } from '../src/ports/runtime.js';

// Workstream E: a summary covers what it summarized, so the prompt stops
// carrying it, a thread compacts once per time it outgrows the trigger, and
// the kept tail starts at a user turn. The same cases run in the Go package
// (compaction_test.go).

/** Answers every streamed call "ok" and every summary request "summary", and
 *  keeps the prompts it streamed. */
function summarizingModel() {
  const prompts: any[] = [];
  const model = new MockLanguageModelV1({
    provider: 'mock',
    modelId: 'mock-summarizing',
    doGenerate: async () => ({
      text: 'summary',
      finishReason: 'stop',
      usage: { promptTokens: 100, completionTokens: 10 },
      rawCall: { rawPrompt: null, rawSettings: {} },
    }),
    doStream: async ({ prompt }: any) => {
      prompts.push(prompt);
      const chunks: LanguageModelV1StreamPart[] = [
        { type: 'text-delta', textDelta: 'ok' },
        { type: 'finish', finishReason: 'stop', usage: { promptTokens: 10, completionTokens: 5 } },
      ];
      return { stream: simulateReadableStream({ chunks }), rawCall: { rawPrompt: null, rawSettings: {} } };
    },
  });
  return { model, lastPrompt: () => JSON.stringify(prompts.at(-1)) };
}

/** A 1,800-token budget, compacting past 1,440, keeping a 450-token tail.
 *  Each prompt below is about 300 tokens. */
async function compactingRuntime() {
  const { model, lastPrompt } = summarizingModel();
  const storage = new MemoryStorage();
  const queue = new MemoryQueue();
  const bus = new MemoryBus();
  const kv = new MemoryKv();
  const admin = new MemoryAdminStore();
  const config = resolveConfig({
    contextCeilingTokens: 2_000,
    contextOutputReserveTokens: 200,
    compactionModel: 'gpt-4o',
    promptCaching: false,
  });
  const resolveModel = () => ({ instance: () => model, contextWindow: 128_000 });
  const runtime = await setupAgentCore({ storage, queue, bus, kv, admin, config, resolveModel });
  const chat = runtime.createStreamTextAgent({ name: 'chat', model: 'gpt-4o' });
  const ports: RuntimePorts = {
    storage: bindStorage(storage, { state: {} }), queue, bus, kv, admin, config, resolveModel,
  };
  const runPrompt = async (threadId: string | undefined, i: number) => {
    const ran = await chat.run({ ...(threadId ? { threadId } : {}), prompt: `run-${i} ${'x'.repeat(1_200)}` });
    await runtime.worker.handleJob(queue.items.shift()!);
    return ran;
  };
  const summaries = (threadId: string) =>
    storage.messages.store.get(threadId)!.filter((m) => (m.content as any)?.type === 'CONTEXT_SUMMARY').length;
  return { storage, ports, lastPrompt, runPrompt, summaries };
}

describe('compaction (§2.6)', () => {
  it('a thread compacts once each time it grows past the trigger', async () => {
    const r = await compactingRuntime();
    const first = await r.runPrompt(undefined, 1);
    for (let i = 2; i <= 6; i++) await r.runPrompt(first.threadId, i);
    expect(r.summaries(first.threadId)).toBe(1); // not on every run after
    expect(r.lastPrompt()).not.toContain('run-1 '); // a summarized turn is not sent again
    expect(r.lastPrompt()).toContain('run-6 ');
  });

  it('the summary is billed to the run it served', async () => {
    const r = await compactingRuntime();
    const first = await r.runPrompt(undefined, 1);
    let compacting = first;
    for (let i = 2; i <= 5; i++) compacting = await r.runPrompt(first.threadId, i);
    const row = r.storage.usage.recorded.find((u) => u.kind === 'compaction');
    expect(row?.runId).toBe(compacting.runId);
  });

  it('the kept tail starts at a user turn', async () => {
    const r = await compactingRuntime();
    const th = await r.storage.threads.create({});
    const add = (role: string, content: unknown) => r.storage.messages.append(th.id, { role: role as any, content });
    const big = 'y'.repeat(2_400);
    await add('user', `first ${big}`);
    await add('user', `second ${big}`);
    await add('assistant', [
      { type: 'tool-call', toolCallId: 'c1', toolName: 'lookup', args: { q: 'q'.repeat(300) } },
    ]);
    // A result that fits the tail on its own, with its call just outside it.
    await add('tool', [{ type: 'tool-result', toolCallId: 'c1', toolName: 'lookup', result: 'z'.repeat(1_300) }]);
    await add('assistant', 'looked it up');
    await add('user', 'third');
    const out = await compactContext(r.ports, th.id, 'gpt-4o');
    expect(out.map((m) => m.role)).toEqual(['system', 'user']); // summary + the last user turn
  });

  it('a summary from before the cover mark is left out', () => {
    const msg = (id: string, role: string, content: unknown) => ({ id, role, content });
    const u1 = msg('u1', 'user', 'one');
    const a1 = msg('a1', 'assistant', 'two');
    const u2 = msg('u2', 'user', 'three');
    const legacy = msg('s0', 'system', { type: 'CONTEXT_SUMMARY', text: 'old' });
    const marked = msg('s1', 'system', { type: 'CONTEXT_SUMMARY', text: 'new', coversUpTo: 'a1' });
    expect(promptHistory([u1, a1, legacy, u2]).map((m) => m.id)).toEqual(['u1', 'a1', 'u2']);
    expect(promptHistory([u1, a1, u2, marked]).map((m) => m.id)).toEqual(['s1', 'u2']);
  });
});
